Quadriga/src/main/java/edu/asu/spring/quadriga/service/impl/passthroughproject/PassThroughProjectManager.java

```diff
@@ -0,0 +1,136 @@
+        QuadrigaUserDTO owner = userDAO.getDTO(user.getUserName());
+        PassThroughProjectDTO projectDTO = projectMapper.getPassThroughProjectDTO(project, user);
+        projectDTO.setProjectid(projectId);
+        projectDTO.setProjectowner(owner);
+
+        projectDao.addPassThroughProject(projectDTO);
+
+        return projectId;
+    }
+
+    private String getInternalProjectId(String externalProjectid, String userid)
+            throws QuadrigaStorageException, NoSuchRoleException {
+
+        List<PassThroughProjectDTO> projectDTOs = projectDao.getExternalProjects(externalProjectid);
+
+        for (PassThroughProjectDTO projectDTO : projectDTOs) {
+            boolean isOwner = projectSecurityChecker.isProjectOwner(userid, projectDTO.getProjectid());
+            if (isOwner) {
+                return projectDTO.getProjectid();
+            }
+            boolean isCollaborator = projectSecurityChecker.isCollaborator(userid,
+                    RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN, projectDTO.getProjectid());
+            if (isCollaborator) {
+                return projectDTO.getProjectid();
+            }
+        }
+        return null;
+    }
+
+    @Transactional
+    @Override
+    public String createWorkspaceForExternalProject(PassThroughProjectInfo passThroughProjectInfo, String projectId,
+            IUser user) throws JAXBException, QuadrigaStorageException, QuadrigaAccessException {
+
+        String externalWorkspaceId = passThroughProjectInfo.getExternalWorkspaceId();
+        String externalWorkspaceName = passThroughProjectInfo.getWorkspaceName();
+
+        boolean isExternalWorkspaceExists = externalWSManager.isExternalWorkspaceExists(externalWorkspaceId);
+        String workspaceId = null;
+        if (!isExternalWorkspaceExists) {
+            // External workspace does not exists so insert the values into
+            // externalWorkspace table
+            // Create a new externalWorkspaceId and InternalWorkspaceId and then
+            // call storeNetworkDetails
+            workspaceId = workspaceDao.generateUniqueID();
+            externalWSManager.createExternalWorkspace(externalWorkspaceId, externalWorkspaceName, workspaceId,
+                    projectId, user);
+        } else {
+            // Get the workspace Id related to the external workspace Id
+            workspaceId = externalWSManager.getInternalWorkspaceId(externalWorkspaceId);
+        }
+
+        // After creating the values save the values by calling store network
+        // details. If already available. Call network details for updation
+
+        return workspaceId;
+
+    }
+
+    @Transactional
+    @Override
+    public String savePassThroughProject(IUser user, IPassThroughProject project)
+            throws QuadrigaStorageException, NoSuchRoleException {
+
+        String internalProjetid = getInternalProjectId(project.getExternalProjectid(), user.getUserName());
```

Comment:
The method should return null if there is no project for the external project id and then you can simply check for null.